numpy/_core/fromnumeric.py

```diff
@@ -206,6 +207,100 @@ def take(a, indices, axis=None, out=None, mode='raise'):
     return _wrapfunc(a, 'take', indices, axis=axis, out=out, mode=mode)
 
 
+def _top_k_dispatcher(a, k, /, *, axis=-1, largest=True):
+    return (a,)
+
+
+@array_function_dispatch(_top_k_dispatcher)
+def top_k(a, k, /, *, axis=-1, largest=True):
+    """
+    Returns the ``k`` largest/smallest elements and corresponding
+    indices along the given ``axis``.
+
+    When ``axis`` is None, a flattened array is used.
+
+    If ``largest`` is false, then the ``k`` smallest elements are returned.
+
+    A tuple of ``(values, indices)`` is returned, where ``values`` and
+    ``indices`` of the largest/smallest elements of each row of the input
+    array in the given ``axis``.
```

Comment:
It would be useful to explicitly note the semantics in the presence of NaN values. Is this the same as `sort(a)[:k]` / `sort(a)[-k:]`, or it the same as `sort(a[~isnan(a)])[:k]` / `sort(a[~isnan(a)])[-k:]`?

Also: does the API make any guarantees about the order of the returned results?

Reply:
With regards to `np.nan`, from what I understand, the underlying `np.argpartition` is not intentional in how it treats `np.nan`. For floats by the nature of how the partial sort is implemented, `np.nan` is treated like `np.inf` since it fails for every comparison with a number. This might change in the future as the underlying implementation changes. Should I add a note that the treatment of `np.nan` is not defined?

About the order of the returned results, `np.argpartition` by default uses a partial sort which is unstable, so the returned indices is not guaranteed to be the first occurrence of the element. E.g., `np.top_k([3,3], 1)` returns `(array([3]), array([1]))`. I'll add that as a note.

Reply:
NumPy uses a sort order pushing NaNs to the end consistently.  I don't think we should change that.

Now, there *is* a problem with respect to adding a kwarg to choose a descending sort (which you propose here for `top_k`).  In that case it might be argued that NaNs should _also_ be sorted to the end!
And if we want that, it would require specific logic to sort in descending order (not just for unstable sorts).

Reply:
When I asked about the order of returned elements, what I had in mind was this:
```python
np.top_k([1, 4, 2, 3], k=2)
```
As I see it, there are three logically-consistent conventions:

- results are always sorted: return `[3, 4]`
- results are always in the order they appear: return `[4, 3]`
- order is not guaranteed: return either `[3, 4]` or `[4, 3]`

It would be helpful to specify in the documentation which of these is the case for NumPy's implementation.

Reply:
It looks like the current implementation returns results that are always sorted. Which seems to me like it's the nicest option for the user. If that falls out naturally, then great. And for other implementations, matching that behavior doesn't seem too costly performance-wise even if their implementation doesn't yield sorted values naturally, because the returned arrays are typically very small compared to the input arrays, so sorting the end result is fast.

Does that sound right to everyone?

@JuliaPoo based on the implementation, do you see a reason that this is hard to guarantee?